### PR TITLE
feat: Mobile Vitals - Native Frames

### PR DESCRIPTION
**Pull request by denrase**
## :scroll: Description

- Sentry navigator will fetch native frames over native channel and append to `SentryTransaction` for transaction created automatically between navigation.
- Introduce `onFinish` callback when starting transactions.
- Introduce `enableAutoPerformanceTracking` flag in flutter options.

- Bump `sentry-android` to 5.6.3

## :bulb: Motivation and Context

Implement native frame tracking of mobile vitals. 

## :green_heart: How did you test it?

- Ran iOS/Android example application.
- Unit tests.

## :pencil: Checklist

- [x] I reviewed submitted code
- [ ] I added tests to verify changes
- [ ] I updated the docs if needed
- [x] All tests passing
- [x] No breaking changes

## :crystal_ball: Next steps

More tests!

### CODE REVIEW

File: flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt
Comment (marandaneto) [inline flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt]: What if we make `ActivityFramesTracker` (see Android repo) public but internal, and rather use it? DRY

Comment (marandaneto) [reply flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt]: https://github.com/getsentry/sentry-java/pull/1931

File: dart/lib/src/integration.dart
Comment (marandaneto) [inline dart/lib/src/integration.dart]: The `Integration` class is part of the Unified API, I think this is not a good idea if we can do it differently.

File: flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt
Comment (marandaneto) [inline flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt]: You can check if `metrics` is null only, otherwise everything else would have a value

File: flutter/ios/Classes/SentryFlutterPluginApple.swift
Comment (marandaneto) [inline flutter/ios/Classes/SentryFlutterPluginApple.swift]: Let's `log` only, that's not a vital feature for the SDK.

File: dart/lib/src/hub.dart
Comment (marandaneto) [inline dart/lib/src/hub.dart]: What's about a `typedef`?

`typedef OnTransactionFinish = FutureOr<void> Function(ISentrySpan transaction);`

File: flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt
Comment (marandaneto) [inline flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt]:
```suggestion
    activity = nul
```

Maybe we clean the reference or make `activity` a WeakRef?
And remove from `activityFramesTracker` as well.

File: flutter/ios/Classes/SentryFlutterPluginApple.swift
Comment (marandaneto) [inline flutter/ios/Classes/SentryFlutterPluginApple.swift]: We should only enable these 2 flags if the performance is enabled on Flutter, Similar to RN.
https://github.com/getsentry/sentry-react-native/blob/main/ios/RNSentry.m#L88-L94

File: flutter/ios/Classes/SentryFlutterPluginApple.swift
Comment (marandaneto) [inline flutter/ios/Classes/SentryFlutterPluginApple.swift]: Let's check if they are not zeroed before returning the map.
https://github.com/getsentry/sentry-react-native/blob/main/ios/RNSentry.m#L226-L234

File: flutter/lib/src/integrations/mobile_vitals_integration.dart
Comment (marandaneto) [inline flutter/lib/src/integrations/mobile_vitals_integration.dart]: `transaction.eventId` is the same id iirc, and it's not nullable.

File: flutter/lib/src/navigation/sentry_navigator_observer.dart
Comment (marandaneto) [inline flutter/lib/src/navigation/sentry_navigator_observer.dart]: What if we also have an `onStart` callback, that's called during the transaction creation, so this can be reused for some other reasons.

File: flutter/lib/src/navigation/sentry_navigator_observer.dart
Comment (marandaneto) [inline flutter/lib/src/navigation/sentry_navigator_observer.dart]: Could we here get the map that comes from the Native SDK and append it to the `transaction` itself already? Instead of caching it somewhere and reading thru an event processor.

We could add a method `addMeasurement` on the `SentryTracer` class.

And here we could do:

```
if (transaction is SentryTracer) {
   // add attachment
}
```

File: flutter/lib/src/navigation/sentry_navigator_observer.dart
Comment (marandaneto) [inline flutter/lib/src/navigation/sentry_navigator_observer.dart]:
```suggestion
  Future<void> _startTransaction(String? name, Object? arguments) async {
```

File: flutter/lib/src/navigation/sentry_navigator_observer.dart
Comment (denrase) [reply flutter/lib/src/navigation/sentry_navigator_observer.dart]: We can't do that without a warning, as `SentryTracer ` is internal to the dart package.

File: flutter/lib/src/navigation/sentry_navigator_observer.dart
Comment (denrase) [reply flutter/lib/src/navigation/sentry_navigator_observer.dart]: Thought about that, but i'm nur sure this is not redundant? As i can always do what i wanted to do after i create the transaction, no need for the callback? WDYT?

File: flutter/lib/src/integrations/mobile_vitals_integration.dart
Comment (denrase) [reply flutter/lib/src/integrations/mobile_vitals_integration.dart]: Yeah, on the other hand when accessing the `ISentryFrame` objects form other places, we always have to access through `contexts.trace?.traceId`, so it makes sense that this reads the same, no?

File: flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt
Comment (denrase) [reply flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt]: Compiler will still assume nullability if we don't check the individual variables, as it we are pulling them out of a dictionary.

File: flutter/lib/src/navigation/sentry_navigator_observer.dart
Comment (ueman) [reply flutter/lib/src/navigation/sentry_navigator_observer.dart]: Yeah, the structure of the SDK is a bit weird. I've noticed that during my attempts at metrics.
All the public API is based on `ISentrySpan`, but metrics can only be added to a transaction (as per docs). However, the real implementation on which the metrics can be added is the `SentryTracer` class. So there's no real class where a `addMetric` method would feel right.

File: flutter/lib/src/integrations/mobile_vitals_integration.dart
Comment (marandaneto) [reply flutter/lib/src/integrations/mobile_vitals_integration.dart]: `contexts.trace` is built out from the `transaction.eventId` anyway, I just thought about it because the eventId is not nullable, but your call here, I'm fine either way, just a nit.

File: flutter/lib/src/navigation/sentry_navigator_observer.dart
Comment (marandaneto) [reply flutter/lib/src/navigation/sentry_navigator_observer.dart]: @ueman you are totally right, but measurements, for now, are internal only, it should not have a public API for doing so, that's why we're having this experience.
We're gonna have a public API soon for that, people are experimenting on other SDKs, and once this lands, we can add it here and improve our integrations as well.

File: flutter/lib/src/integrations/mobile_vitals_integration.dart
Comment (denrase) [reply flutter/lib/src/integrations/mobile_vitals_integration.dart]: Obsolete, as we are now using `SentryTracer` instead.

File: dart/lib/src/sentry_tracer.dart
Comment (marandaneto) [inline dart/lib/src/sentry_tracer.dart]:
```suggestion
  void addMeasurements(List<SentryMeasurement> measurements) {
```

File: dart/lib/src/sentry_tracer.dart
Comment (marandaneto) [inline dart/lib/src/sentry_tracer.dart]: Instead of passing via ctor, I'd rather call `transaction.addMeasurements(...)`, the reason is, we'd not need an optional ctor param for `SentryTransaction` only for that.

File: flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt
Comment (marandaneto) [inline flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt]: We could do that after reading `enableAutoPerformanceTracking` from options, so we know its already created if `enableAutoPerformanceTracking` is enabled (nit).

File: flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt
Comment (marandaneto) [inline flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt]:
```suggestion
      Log.w("Sentry", "Parameter id cannot be null when calling endNativeFrames.")
```

File: flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt
Comment (marandaneto) [inline flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt]:
```suggestion
    val total = metrics?.get("frames_total")?.getValue()?.toInt() ?: 0
```

File: flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt
Comment (marandaneto) [inline flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt]: By doing: https://github.com/getsentry/sentry-dart/pull/772/files#r825450103
We can avoid this check.

File: flutter/lib/src/event_processor/native_app_start_event_processor.dart
Comment (marandaneto) [inline flutter/lib/src/event_processor/native_app_start_event_processor.dart:19]: What if we don't use this EventProcessor anymore and do it within the `onFinish` as well? Does it make sense?

**Comment by marandaneto** [comment]
@denrase Just left a few minor comments, but the overall feature looks good, thanks for doing this.

File: flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt
Comment (brustolin) [inline flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt:81]: l: Are we required to override this function even though we have no code to add? Can we just delete this empty functions?

File: flutter/lib/src/sentry_native.dart
Comment (brustolin) [inline flutter/lib/src/sentry_native.dart]: l: This for me is a "code smell". Having something that looks like a constructor returning a shared instance, and you can even reconfigure it, which would affect it everywhere. 
I would rather have a static "shared" property.
I understand that dart has this concept of factory constructors, and singletons is one of the examples `factory` documentation suggests. But, only because we can, doesn't mean we should.

File: flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt
Comment (marandaneto) [reply flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt:81]: Unless methods have a default signature, you have to override.

File: flutter/lib/src/sentry_native.dart
Comment (marandaneto) [reply flutter/lib/src/sentry_native.dart]: GDEs for Flutter recommend this approach for Singletons, See https://twitter.com/vandadnp/status/1502860065270808582

The question for me here is, does it need to be a singleton? Looks like we can pass `SentryNative` instances via ctor to the needed integrations.

File: flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt
Comment (brustolin) [reply flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt:81]: Ow, because we are implementing an interface and not inheriting a class, got it. 
Im spending too much time with objc, lol.

File: flutter/lib/src/sentry_native.dart
Comment (brustolin) [reply flutter/lib/src/sentry_native.dart]: > GDEs for Flutter recommend this approach for Singletons, See

I understand that. I just don't think this is a good thing to do. It's easier to make mistakes this way because the dev could think is getting a new object instead of an instance. 

But anyway, is just a personal preference. 

>The question for me here is, does it need to be a singleton? Looks like we can pass SentryNative instances via ctor to the needed integrations.

IMO it make sense to have classes that deal only with channel communication as singleton.

Comment (denrase) [reply flutter/lib/src/sentry_native.dart]: > The question for me here is, does it need to be a singleton? Looks like we can pass SentryNative instances via ctor to the needed integrations.

We need some shared state, because we hold properties (e.g. `appStartEnd`) in this class, and we don't really have any other way to hold onto them except for options.

> But anyway, is just a personal preference.

This was also my take, I'm not super happy about having a singleton returned with `Class()` syntax, but also always preferred something like `Class.sharedInstance`, because you immediately see what it is. But if that's the way to go in `dart` land and we decide to do it consistently throughout the SDK, we should stick with it.

As for the `ctor` parameter, in essence I needed a way to provide the `Channel`, which is used internally, at a later time than instance creation, as it is only available to us when users call `init`.

We could at least move this injection to a method to make it more explicit. WDYT?

File: flutter/lib/src/event_processor/native_app_start_event_processor.dart
Comment (denrase) [reply flutter/lib/src/event_processor/native_app_start_event_processor.dart:19]: Think this was one of the first feedback we had to move it away from the `SentryNavigatorObserve` so that users who don't init the observer would also have this feature.

Comment (marandaneto) [reply flutter/lib/src/event_processor/native_app_start_event_processor.dart:19]: Indeed, makes total sense.

File: flutter/lib/src/sentry_native.dart
Comment (marandaneto) [reply flutter/lib/src/sentry_native.dart]: `SentryNative` needs to be used in the `SentryNavigatorObserver`, and `SentryNavigatorObserver` is created by the user, so it needs to be a singleton/shared instance.
Another way would be if `SentryNavigatorObserver` has access to the MethodChannel, then it could pass the instance per method call for `SentryNative`, but not sure if it makes much difference.
@denrase your call.

File: dart/lib/src/sentry_tracer.dart
Comment (denrase) [reply dart/lib/src/sentry_tracer.dart:111]: If going down that route we always have to init the instance, since the class is marked as `@immutable` and we'd also have do do additional work when calling the copy method.

Comment (denrase) [reply dart/lib/src/sentry_tracer.dart]: We do need it as events are copied when going through processing and we'd loose measurements otherwise.

File: flutter/lib/src/sentry_native.dart
Comment (denrase) [reply flutter/lib/src/sentry_native.dart]: @marandaneto Then we'd somehow need to provide the channel instance to the SentryNavigatorObserver, but it is a static in `SentryFlutter` and can also be overridden in the init method.

@brustolin We now use method injection for the channel dependency, so the intent is clear and there's no unneeded additional confusion with the `ctor` call.

File: CHANGELOG.md
Comment (marandaneto) [inline CHANGELOG.md]:
```suggestion
```

File: flutter/android/build.gradle
Comment (marandaneto) [inline flutter/android/build.gradle]:
```suggestion
    api 'io.sentry:sentry-android:5.6.3'
```

File: flutter/lib/src/sentry_flutter.dart
Comment (marandaneto) [inline flutter/lib/src/sentry_flutter.dart]: I see we are passing `native` as param in a few methods, do we need it? since it's a Singleton.

**Review by marandaneto — APPROVED** [review]
@denrase LGTM, left just a comment tho, https://github.com/getsentry/sentry-dart/pull/772/files#r828893553
@brustolin please double-check :)

File: CHANGELOG.md
Comment (marandaneto) [inline CHANGELOG.md]:
```suggestion
## Unreleased
```

File: flutter/ios/Classes/SentryFlutterPluginApple.swift
Comment (brustolin) [inline flutter/ios/Classes/SentryFlutterPluginApple.swift]: picking: 
```suggestion
            if arguments["enableAutoPerformanceTracking"] as? Bool ?? false {
```

File: flutter/ios/Classes/SentryFlutterPluginApple.swift
Comment (marandaneto) [inline flutter/ios/Classes/SentryFlutterPluginApple.swift]: We actually need to check if there's UI Kit and not iOS only.
https://github.com/getsentry/sentry-cocoa/blob/17969fa1c80d643df62aa92e8a5fa724c9af0d8c/Sources/Sentry/SentryFramesTrackingIntegration.m#L22-L36

https://github.com/getsentry/sentry-cocoa/blob/f893682911568e14782eb0bb507af5a12332b033/Sources/Sentry/Public/SentryDefines.h#L9-L19

Does it work if we do `os(iOS) || os(whatElseWorks`)?

Comment (marandaneto) [reply flutter/ios/Classes/SentryFlutterPluginApple.swift]: @philipphofmann Can you help us out here?

Comment (marandaneto) [reply flutter/ios/Classes/SentryFlutterPluginApple.swift]: It should work for:

> This feature is available for iOS, tvOS, and Mac Catalyst.

**Review by brustolin — APPROVED** [review]
LGTM

File: flutter/ios/Classes/SentryFlutterPluginApple.swift
Comment (brustolin) [reply flutter/ios/Classes/SentryFlutterPluginApple.swift]: > Does it work if we do os(iOS) || os(whatElseWorks)?

Yes it does

```suggestion
               #if os(iOS) || os(tvOS) || targetEnvironment(macCatalyst)
```

Comment (brustolin) [reply flutter/ios/Classes/SentryFlutterPluginApple.swift]: As far as I know Flutter does not work with tvOS.

Comment (philipphofmann) [reply flutter/ios/Classes/SentryFlutterPluginApple.swift]: 👍 to what @brustolin wrote.

Comment (marandaneto) [reply flutter/ios/Classes/SentryFlutterPluginApple.swift]: Indeed, no tvOS for now, https://docs.flutter.dev/deployment/macos

File: flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt
Comment (marandaneto) [inline flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt]: @denrase '@file:' annotations can only be applied before package declaration

Comment (denrase) [reply flutter/android/src/main/kotlin/io/sentry/flutter/SentryFlutterPlugin.kt:30]: Interesting, this did work for locally executing `detekt` with gradle.

**Pull request merged**
